File: day-8/go/part2.go
package day8

import (
	"fmt"
	"os"
	"strings"
	"time"
)

func (d Day8) Part2(filename *string) string {
	// start clock
	start := time.Now()

	content, _ := os.ReadFile(*filename)
	lines := strings.Split(string(content), "\n")

	var network = map[string]Direction{}
	var instructions string
	var starting_elements []string

	// create network in form of map
	for idx, line := range lines {
		if idx == 1 {
			continue
		}

		if idx == 0 {
			instructions = line
			continue
		}

		var parts = strings.Split(line, " = ")
		var direction = strings.Split(parts[1], ", ")
		var left_direction = direction[0][1:]
		var right_direction = direction[1][:3]
		network[parts[0]] = Direction{left_direction, right_direction}

		if parts[0][2] == 'A' {
			starting_elements = append(starting_elements, parts[0])
		}
	}

	var steps_to_z []int

	// For each starting element find the number of steps to a Z
	// and add it to the list
	for _, starting_element := range starting_elements {
		// Start at the first element of the network
		var steps = 0
		var curr_element = fmt.Sprint(starting_element)
		var instruction_idx = 0

		// Keep going until we reach a Z
		for curr_element[len(curr_element)-1] != 'Z' {
			var directions = network[curr_element]
			var curr_direction = instructions[instruction_idx]

			// Choose a direction
			if string(curr_direction) == "L" {
				curr_element = directions.Left
			} else {
				curr_element = directions.Right
			}

			// Make sure the instruction index doesn't go out of bounds
			if instruction_idx == len(instructions)-1 {
				instruction_idx = 0
			} else {
				instruction_idx++
			}

			// Increment the number of steps
			steps++
		}

		steps_to_z = append(steps_to_z, steps)
	}

	// To find the number of steps to Z where all the starting numbers
	// have the same number of steps to Z, we need to find the LCM of
	// all the steps to Z
	var steps = LCM(steps_to_z[0], steps_to_z[1], steps_to_z[2:]...)

	fmt.Println(time.Since(start))
	return fmt.Sprint(steps)
}

/**
 * The following two functions are taken from the GO.dev website
 * https://go.dev/play/p/SmzvkDjYlb
 */
// greatest common divisor (GCD) via Euclidean algorithm
func GCD(a, b int) int {
	for b != 0 {
		t := b
		b = a % b
		a = t
	}
	return a
}

// find Least Common Multiple (LCM) via GCD
func LCM(a, b int, integers ...int) int {
	result := a * b / GCD(a, b)

	for i := 0; i < len(integers); i++ {
		result = LCM(result, integers[i])
	}

	return result
}

File: day-08/cpp/part2.cpp
// To run do: c++ part2.cpp -Wall -std=c++17 -o output2.out && ./output2.out
#include <string>
#include <vector>
#include <fstream>
#include <iostream>
#include <map>
#include "../../utils/utils.cpp"

using namespace std;

// greatest common divisor (GCD) via Euclidian algorithm
long GCD(long a, long b)
{
    while (b != 0)
    {
        long t = b;
        b = a % b;
        a = t;
    }

    return a;
}

// least common divisor (LCM) via GCD
long LCM(vector<long> integers)
{
    long result = integers[0] * integers[1] / GCD(integers[0], integers[1]);

    for (long i = 2; i < integers.size(); i++)
    {
        result = result * integers[i] / GCD(result, integers[i]);
    }

    return result;
}

int main()
{
    // Initialize a vector to store the file input and set the input file name.
    std::string fileName = "../input.txt";
    std::vector<std::string> lines = read_file(fileName);

    map<string, tuple<string, string>> network;
    string instructions;
    vector<string> starting_elements;

    for (int i = 0; i < lines.size(); i++)
    {
        if (i == 1)
        {
            continue;
        }

        if (i == 0)
        {
            instructions = lines[i];
            continue;
        }

        vector<string> parts = splitter(lines[i], " = ");
        vector<string> direction = splitter(parts[1], ", ");
        string left_direction = splitter(direction[0], "(")[1];
        string right_direction = splitter(direction[1], ")")[0];
        network[parts[0]] = tuple(left_direction, right_direction);

        if (parts[0][2] == 'A')
        {
            starting_elements.push_back(parts[0]);
        }
    }

    vector<long> steps_to_z;

    for (long i = 0; i < starting_elements.size(); i++)
    {
        string curr_element = starting_elements[i];
        long instruction_idx = 0;
        long steps = 0;
        while (curr_element[2] != 'Z')
        {
            tuple<string, string> directions = network[curr_element];
            char instruction = instructions[instruction_idx];

            if (instruction == 'L')
            {
                curr_element = get<0>(directions);
            }
            else
            {
                curr_element = get<1>(directions);
            }

            if (instruction_idx == instructions.length() - 1)
            {
                instruction_idx = 0;
            }
            else
            {
                instruction_idx += 1;
            }

            steps += 1;
        }

        steps_to_z.push_back(steps);
    }

    // Print the result.
    std::cout << "Result: " << LCM(steps_to_z) << std::endl;
}